test/decl/func/dynamic_self.swift

```diff
@@ -412,4 +412,88 @@ class SelfOperator {
 func useSelfOperator() {
   let s = SelfOperator()
   _ = s + s
-}
\ No newline at end of file
+}
+
+// These references to Self are now possible
+
+class A<T> {
+  let b: Int
+  required init(a: Int) {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    b = a
+  }
+  static func z() {
+    print("\(Self.self).\(#function)")
+  }
+  class func y() {
+    print("\(Self.self).\(#function)")
+    Self.z()
+  }
+  func x() -> A? {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    Self.z()
+    return Self.init(a: 77)
+  }
+}
+
+class B: A<Int> {
+  let a: Int
+  required convenience init(a: Int) {
+    print("\(Self.self).\(#function)")
+    self.init()
+  }
+  init() {
+    print("\(Self.self).\(#function)")
+    Self.y()
+    Self.z()
+    a = 99
+    super.init(a: 88)
+  }
+  override class func y() {
+    print("override \(Self.self).\(#function)")
+  }
+}
+
+struct S2 {
+  let x = 99
+  static func x() {
+    Self.y()
+  }
+  static func y() {
+    print("y()")
+  }
+  func foo(a: [Self]) -> Self? {
+    Self.x()
+    return Self.init() as Self
+  }
+}
+
+struct S3<T> {
+  let x = 99
+  static func x() {
+    Self.y()
+  }
+  static func y() {
+    print("y()")
+  }
+  func foo(a: [Self]) -> Self? {
+    Self.x()
+    return Self.init() as? Self
```

Comment:
Add a test case that does this in a class as well, both with and without the 'as? Self'

Reply:
This tis tested in the generic class version as well now.